node/core/pvf/src/execute/worker.rs

```diff
@@ -0,0 +1,233 @@
+		return Outcome::IoErr;
+	}
+
+	let response = futures::select! {
+		response = recv_response(&mut stream).fuse() => {
+			match response {
+				Err(_err) => return Outcome::IoErr,
+				Ok(response) => response,
+			}
+		},
+		_ = Delay::new(Duration::from_secs(3)).fuse() => return Outcome::HardTimeout,
+	};
+
+	match response {
+		Response::Ok {
+			result_descriptor,
+			duration_ms,
+		} => Outcome::Ok {
+			result_descriptor,
+			duration_ms,
+			idle_worker: IdleWorker { stream, pid },
+		},
+		Response::InvalidCandidate(err) => Outcome::InvalidCandidate {
+			err,
+			idle_worker: IdleWorker { stream, pid },
+		},
+	}
+}
+
+async fn send_request(
+	stream: &mut UnixStream,
+	artifact_path: &Path,
+	validation_params: &[u8],
+) -> io::Result<()> {
+	framed_send(stream, path_to_bytes(artifact_path)).await?;
+	framed_send(stream, validation_params).await?;
+	Ok(())
+}
+
+async fn recv_request(stream: &mut UnixStream) -> io::Result<(PathBuf, Vec<u8>)> {
+	let artifact_path = framed_recv(stream).await?;
+	let artifact_path = bytes_to_path(&artifact_path).ok_or_else(|| {
+		io::Error::new(
+			io::ErrorKind::Other,
+			"execute pvf recv_request: non utf-8 artifact path".to_string(),
+		)
+	})?;
+	let params = framed_recv(stream).await?;
+	Ok((artifact_path, params))
+}
+
+async fn send_response(stream: &mut UnixStream, response: Response) -> io::Result<()> {
+	framed_send(stream, &response.encode()).await?;
+	Ok(())
+}
+
+async fn recv_response(stream: &mut UnixStream) -> io::Result<Response> {
+	let response_bytes = framed_recv(stream).await?;
+	let response = Response::decode(&mut &response_bytes[..]).map_err(|e| {
+		io::Error::new(
+			io::ErrorKind::Other,
+			format!("execute pvf recv_response: decode error: {:?}", e),
+		)
+	})?;
+	Ok(response)
+}
+
+#[derive(Encode, Decode)]
+enum Response {
+	Ok {
+		result_descriptor: ValidationResult,
+		duration_ms: u64,
+	},
+	InvalidCandidate(String),
+}
+
+impl Response {
+	fn format_invalid(ctx: &'static str, msg: &str) -> Self {
+		if msg.is_empty() {
+			Self::InvalidCandidate(ctx.to_string())
+		} else {
+			Self::InvalidCandidate(format!("{}: {}", ctx, msg))
+		}
+	}
+}
+
+/// The entrypoint that the spawned execute worker should start with. The socket_path specifies
+/// the path to the socket used to communicate with the host.
+pub fn worker_entrypoint(socket_path: &str) {
+	worker_event_loop("execute", socket_path, |mut stream| async move {
+		let executor = TaskExecutor::new().map_err(|e| {
+			io::Error::new(
+				io::ErrorKind::Other,
+				format!("cannot create task executor: {}", e),
+			)
+		})?;
+		loop {
+			let (artifact_path, params) = recv_request(&mut stream).await?;
+			let artifact_bytes = async_std::fs::read(&artifact_path).await?;
+			let artifact = Artifact::deserialize(&artifact_bytes).map_err(|e| {
+				io::Error::new(
+					io::ErrorKind::Other,
+					format!("artifact deserialization error: {}", e),
+				)
+			})?;
```

Comment:
Is it worth sending a response of some kind even if this fails, or will the host learn about the failure some other way?